### ISSUE TITLE
[rllib] Introduce dynamic time-limits into training

### ISSUE THREAD

**Issue by roireshef** [ray-project/ray#12490]
### Describe your feature request
Some environments that can run for long periods are often not advised to do so to generate too long episodes that bias training. Instead, one would often like to artificially limit the length of episodes generated by marking an episode as Done but still bootstrap from the value function to generate a reward approximation. This can't happen within the environment implementation since it is not exposed to the policy and can't sample its value approximation.

see: https://arxiv.org/pdf/1712.00378.pdf for better motivation

At the moment, the only flexibility that exists is the _done_ returned signal, which is binary. In the time-out case, _done_ should definitely be True, but this ultimately means the episode terminates with the _reward_ signal returned by the environment, which is insufficient (value bootstrapping is advised).

I've previously implemented and tested this to successfully work and train with A3C (with no GAE), by introducing a _timeout_ signal into the _info_ dictionary returned by the environment, capturing it by the policy utility functions to generate the appropriate _advantage_. 

Finally, I'm wondering if this is better implemented in a different way already. If not, I suggest adding this as a feature to RLlib. I could open a PR for that, but I'll be waiting for some feedback to understand if this is the suggested practice or not.

**Comment by ericl** [comment]
Have you tried using the env settings?
```
    # === Environment Settings ===
    # Number of steps after which the episode is forced to terminate. Defaults
    # to `env.spec.max_episode_steps` (if present) for Gym envs.
    "horizon": None,
    # Calculate rewards but don't reset the environment when the horizon is
    # hit. This allows value estimation and RNN state to span across logical
    # episodes denoted by horizon. This only has an effect if horizon != inf.
    "soft_horizon": False,
    # Don't set 'done' at the end of the episode. Note that you still need to
    # set this if soft_horizon=True, unless your env is actually running
    # forever without returning done=True.
    "no_done_at_end": False,
```

It sounds like setting horizon and no_done_at_end might do the trick?

**Comment by roireshef** [comment]
@ericl actually I think I'm describing a different use case than what your proposal addresses. In my use case, the episode doesn't only terminate after N steps, but more importantly it terminates *dynamically based on conditions when they are met*. Just to illustrate, consider the following case:
* The environment has an autonomous vehicle agent acting in it. 
1. When the vehicle crashes, _Done_ signal returns True and _Reward_ returns some penalty.
2. When the vehicle succeeds in its mission, _Done_ is again True and _Reward_ returns some positive value
BUT,
3. While it's driving the maximal length of episode should be capped at N, but the environment can also terminate at the point where the agent decided to come to a full stop (even before reaching N steps, which is an instance of what I called "dynamic conditions"). 

In 1-2 the simple API works, but for the two sub-cases of 3 you might want to bootstrap the value function to compute your loss, because the original problem is an "infinite horizon" one (let's assume that). The motivation is this - in reality the AV agent is allowed to come to a full stop, but while training - you wouldn't like to generate a bias in your data where a high percentage of the experiences are of a full stop (same state, same action), but rather you'd like to create high variance and coverage of cases in your data used for training.

In 3 - when the fixed horizon is reached, this is basically already handled by the current API quite OOTB, but there's no easy option for the "dynamic horizon" approach I just detailed. In previous Ray versions I used the _info_ dictionary to carry another boolean signal that indicated whether to bootstrap the value function or not, which worked great, and allowed teaching the policy to break for long time when necessary, without actually creating the bias in training data. But this was prior to the new View API, that saves time by removing boilerplate from the communication, specifically by also not sending the _info_ dict to the loss functions. Not it looks more involved. 

So I'm wondering if this is something you ever thought of, and whether I'm missing already existing solution...?

Does it make more sense now?

**Comment by ericl** [comment]
Hmm if this is about a regression in the traj view API, I think that would be a bug. @sven1977 do you know why the info dict is no longer sent automatically?

**Comment by sven1977** [comment]
Yes, there is a known bug related to the trajectory view API in 1.0.1 described here:
https://github.com/ray-project/ray/issues/12509
The workaround would be to set "_use_trajectory_view_api" to False here (and you should see the info dict again). This was fixed in all versions >1.0.1.

I'll add this observation (missing info dict) to the issue.

**Comment by roireshef** [comment]
@sven1977 @ericl - Is there any documentation on how to use/customize the _trajectory view api_ ? 
I'd like to add this flag that is passed from my environment in the _info_ dict it returns so that _add_advantages_ will have an access to it. It's actually independent of the algo being used.

I understand the easiest way would be to disable the _trajectory view api_ but AFAIK it's better to use this api to achieve better runtimes.

**Comment by sven1977** [comment]
Hey @roireshef , sorry for the delays here. The documentation PR will go into another round of review today: 
https://github.com/ray-project/ray/pull/12718

Again, not seeing the `info` dict in your loss function/postprocessing fn is simply a bug, which can be alleviated by running RLlib w/o the trajectory view API. In >=1.1.x, this should be fixed.
